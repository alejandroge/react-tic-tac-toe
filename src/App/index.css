:root {
  --board-square-size: 50px;
}

body {
  font: 14px "Century Gothic", Futura, sans-serif;
  margin: 20px;
}

ol, ul {
  padding-left: 30px;
}

.board-row:after {
  clear: both;
  content: "";
  display: table;
}

.status {
  margin-bottom: 10px;
}

.square {
  background: #fff;
  border: 1px solid #999;
  float: left;
  font-size: 24px;
  font-weight: bold;
  width: var(--board-square-size);
  height: var(--board-square-size);
  line-height: var(--board-square-size);
  margin-right: -1px;
  margin-top: -1px;
  padding: 0;
  text-align: center;
}

.square:focus {
  outline: none;
}

.kbd-navigation .square:focus {
  background: #ddd;
}

.game {
  display: flex;
  flex-direction: row;
}

.game-info {
  margin-left: 20px;
}

.game-info li {
  list-style: none;
  margin: 5px;
}

.game-info li button {
  width: 150px;
}
.game-info .current-step button {
  font-weight: bold;
}

.game-info .game-status {
  height: var(--board-square-size);
  line-height: var(--board-square-size);
  font-size: 40px;
}

.game-info .moves-list {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  height: calc(var(--board-square-size) * 2);
}
