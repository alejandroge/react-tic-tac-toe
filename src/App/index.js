import React from 'react';
import Board from './Board/'
import './index.css';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      history: [{
        squares: Array(9).fill(null),
        squareNo: null,
      }],
      stepNumber: 0,
      xIsNext: true,
    };
  }

  handleClick(i) {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const current = history[history.length - 1];
    const squares = current.squares.slice();
    if (calculateWinner(squares) || squares[i]) {
      return;
    }
    squares[i] = this.state.xIsNext ? 'X' : 'O';
    this.setState({
      history: history.concat([{
        squares, squareNo: i,
      }]),
      stepNumber: history.length,
      xIsNext: !this.state.xIsNext,
    });
  }

  jumpTo(step) {
    this.setState({
      stepNumber: step,
      xIsNext: (step % 2) === 0,
    });
  }

  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    const moves = history.map((step, move) => {
      let moveDesc = ''
      const instruction = move ?  `Go to move #${move}` : 'Go to game start';
      if (step.squareNo != null) {
        const row = Math.floor(step.squareNo / 3) + 1
        const column = step.squareNo % 3 + 1
        moveDesc = `(${row}, ${column})`
      }
      const klass = (move === this.state.stepNumber) ? 'current-step' : ''
      const desc = [instruction, moveDesc].join(' ')
      return (
        <li key={ move } className={ klass }>
          <button onClick={() => this.jumpTo(move)}>{ desc }</button>
        </li>
      );
    });

    let status
    if (winner) {
      status = `Winner: ${winner}`
    } else {
      status = `Next player: ${this.state.xIsNext ? 'X' : 'O'}`;
    }

    return (
      <div className="game">
        <div className="game-board">
          <Board
            squares={current.squares}
            onClick={(i) => this.handleClick(i)}
          />
        </div>
        <div className="game-info">
          <div className='game-status'>{ status }</div>
          <div className='moves-list'>{ moves }</div>
        </div>
      </div>
    );
  }
}

function calculateWinner(squares) {
  const winningCombos = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6]
  ];

  for(let i = 0; i < winningCombos.length; i++) {
    const [a, b, c] = winningCombos[i];
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
}

export default App;
