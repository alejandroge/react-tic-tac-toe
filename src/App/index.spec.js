import React from 'react';
import renderer from 'react-test-renderer';
import App from './index'

test('html layout is renderer correctly', () => {
  const component = renderer.create(
    <App />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
