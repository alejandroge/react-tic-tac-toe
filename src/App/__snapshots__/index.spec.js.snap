// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html layout is renderer correctly 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div
      className="game-status"
    >
      Next player: X
    </div>
    <div
      className="moves-list"
    >
      <li
        className="current-step"
      >
        <button
          onClick={[Function]}
        >
          Go to game start 
        </button>
      </li>
    </div>
  </div>
</div>
`;
